### PR TITLE
ARROW-5878: [C++][Parquet] Restore pre-0.14.0 Parquet forward compatibility by adding option to unconditionally set TIMESTAMP_MICROS/TIMESTAMP_MILLIS ConvertedType

### PR DESCRIPTION
**Pull request by bkietz**
Set TIMESTAMP_MICROS/TIMESTAMP_MILLIS ConvertedType whether `isAdjustedToUTC` is set in the LogicalType or not

### CODE REVIEW

**Comment by wesm** [comment]
Hmm. So the key changes needed are

* set TIMESTAMP_MICROS/MILLIS unconditional on whether `isAdjustedToUTC` is set. 
* Having this converted type should yield a time zone naive timestamp unless the LogicalType is set. 

This change appears to do more than this. I don't think that TIMESTAMP_MICROS ConvertedType being set should imply `arrow::timestamp(TimeUnit::MICROS, "UTC")`

**Comment by bkietz** [comment]
In that case, is it an error that `LogicalType::FromConvertedType` always sets `isAdjustedToUTC`?

**Comment by bkietz** [comment]
https://github.com/apache/arrow/blob/95d3e06/cpp/src/parquet/types.cc#L383

**Comment by wesm** [comment]
Yes, that should be changed

**Comment by bkietz** [comment]
alright, stand by

**Comment by tpboudreau** [comment]
I understand your motivations, but writing converted type TIMESTAMP_* next to LogicalType(isAdjusted=false) for the same field is inconsistent with the parquet.thrift spec and could confuse a reasonable interpreter (e.g. another library writer).  I believe the spec writers erred in (apparently) layering the semantic nuance about UTC adjustment onto the pre-existing TIMESTAMP_* types after-the-fact.  Whether or not this change proceeds, it might be a good idea to amend the parquet.thrift spec to the effect that TIMESTAMP converted types are UTC-adjustment agnostic, i.e., they apply to both adjusted and non-adjusted LogicalType timestamps.

**Comment by wesm** [comment]
I agree, let's bring it up on the Parquet mailing list. In the meantime, breaking forward compatibility like this for Arrow users is simply not an option. This is the lesser of two evils IMHO

**Comment by wesm** [comment]
Some things in this diff still don't look right to me. I'll take a closer look at this tomorrow when I'm less tired

**Comment by wesm** [comment]
OK I looked. The only fix required is to remove the `if (adjusted)` from here

https://github.com/apache/arrow/blob/master/cpp/src/parquet/types.cc#L1232

The other changes outside of adding appropriate test coverage to ensure the ConvertedType is set in both true/false cases should be reverted

**Comment by tpboudreau** [comment]
> I agree, let's bring it up on the Parquet mailing list. In the meantime, breaking forward compatibility like this for Arrow users is simply not an option. This is the lesser of two evils IMHO

No argument here.  I'll post something on the Parquet list after doing a refresh on the spec.

**Comment by bkietz** [comment]
@wesm I think [similar conditions in LogicalType::Impl::Timestamp::is_compatible()](https://github.com/apache/arrow/blob/master/cpp/src/parquet/types.cc#L1222) need to be changed as well

File: cpp/src/parquet/schema-test.cc [hunk 1]
```diff
@@ -1911,13 +1911,13 @@ TEST_F(TestTemporalSchemaElementConstruction, TemporalCases) {
 
   std::vector<SchemaElementConstructionArguments> timestamp_cases = {
       {"timestamp_T_ms", LogicalType::Timestamp(true, LogicalType::TimeUnit::MILLIS),
-       Type::INT64, -1, true, ConvertedType::TIMESTAMP_MILLIS, true, check_TIMESTAMP},
-      {"timestamp_F_ms", LogicalType::Timestamp(false, LogicalType::TimeUnit::MILLIS),
        Type::INT64, -1, false, ConvertedType::NA, true, check_TIMESTAMP},
+      {"timestamp_F_ms", LogicalType::Timestamp(false, LogicalType::TimeUnit::MILLIS),
+       Type::INT64, -1, true, ConvertedType::TIMESTAMP_MILLIS, true, check_TIMESTAMP},
```
Comment (bkietz) [inline cpp/src/parquet/schema-test.cc]: These tests verify that `ConvertedType::TIMESTAMP_*` is set regardless of timezone awareness

**Comment by wesm** [comment]
I updated the PR title to clarify

**Comment by bkietz** [comment]
Travis is green: https://travis-ci.org/bkietz/arrow/builds/556326351

**Comment by wesm** [comment]
I posted on the Parquet mailing list, I'd like to see if there's any feedback before merging

https://lists.apache.org/thread.html/04b9a8a156939adae1aa4cf2d4e3abd2137f541cbf92ade6d775dc06@%3Cdev.parquet.apache.org%3E

Maybe we can wait 24 hours or so

**Comment by jorisvandenbossche** [comment]
@bkietz also in the reading part, a similar change will be needed (see https://issues.apache.org/jira/browse/ARROW-5889). Do you want to include that here, or do we keep that for a separate PR?

**Comment by wesm** [comment]
@jorisvandenbossche no, let's handle that separately. This is a huge mess since the new LogicalType is applying UTC normalization semantics to (IMHO) data that never had it.

**Review by wesm — APPROVED** [review]
+1. It seems there is sufficient consensus on the mailing list to merge this. I just rebased so we can hopefully get a passing build

**Comment by bkietz** [comment]
@wesm Travis looks green except for a (probably) spurious flight failure https://travis-ci.org/bkietz/arrow/builds/557387598

**Comment by tpboudreau** [comment]
@wesm @bkietz :I'm sorry to jump in fairly late on this, but I think that there might be another option for fixing this with a different set of tradeoffs.  I haven't been able to consider all the usage permutations yet though.

If my PR https://github.com/apache/arrow/pull/4856 goes forward, then it essentially creates a user opt-out from the new TimestampLogicalType UTC inference and causes older files (that present only converted types) to opt-out by default.  As of 0.14.0, Arrow (as a Parquet user) opts-in to the TimestampLogicalType and all its semantics (as it does for all type annotations).  I believe that, rather than changing the parquet-cpp semantics as this patch does, we could have Arrow opt-out by creating timestamp nodes using the legacy constructors that accept converted types.

If I'm right, then the big benefit would be that parquet-cpp implementation would have greatly reduced non-conformance with the spec and the Arrow fixes would be localized to the Arrow sub-libary.  Parquet-cpp would be completely forward compatibility compliant, and it would be backward compatibility compliant, with the proviso that it allows the user to choose whether UTC coercion is desired and does not presume and propagate any choice on that.

The downsides would be that (a) Arrow would be using the deprecated converted types for many timestamps, with no current path forward (until something happens with parquet.thrift spec); and (b) the behavior of Arrow millis and micros timestamps would differ from Arrow nanos timestamps (the former would always be timezone naive after Parquet storage, while the latter would retain their UTC awareness property. 

Should I continue to think about how this might be implemented?

Again, very sorry for the late volley.

**Comment by jorisvandenbossche** [comment]
> The downsides would be that (a) Arrow would be using the deprecated converted types for many timestamps, with no current path forward (until something happens with parquet.thrift spec);

That seems like a considerable downside, as such an option really needs an addition to the parquet.thrift spec in the long term (as we don't want to keep this behaviour indefinitely), and such a change doesn't seem like a given based on the mailing list discussion. 
Also, by doing that, we don't actually enable the new functionality so that drivers that understand the new logical types would still not interpret the arrow written file correctly as timezone-naive.

**Comment by wesm** [comment]
I'm sorry, I'm kind of in combat-medic mode right now with regards to fixing this broken 0.14.0 release -- I would like to fix this regression and move on to other projects that I ought to be working on instead. 

Do you want to move this discussion to the mailing list and present the various options on the table so we can make a decision and move forward?

**Comment by wesm** [comment]
From my point of view, the only acceptable way forward is for Arrow to always set TIMESTAMP_* ConvertedType. This can be behavior that Arrow opts in to.

I don't see a problem with _also_ setting TimestampLogicalType to the semantically correct value. I don't think we should omit the logical type

**Comment by tpboudreau** [comment]
OK, I withdraw the proposal.  Thanks for the feedback.

**Comment by wesm** [comment]
I'm going to pull down this patch and try to make this behavior Arrow-only.

**Comment by wesm** [comment]
@tpboudreau @bkietz take a look, let me know how this seems. The tests show that ConvertedType is not set if using the low level API directly, but going from Arrow sets the option to force setting it

**Comment by tpboudreau** [comment]
Yes.  This feels like a much more elegant approach.  The code looks great.

My only minor comment: do you think the new force_set_converted_type_  property should be included in the ToString and ToJSON output for the TimestampLogicalType?  In my backward compatibility PR -- a bookend to this one where I took essentially the same tack of adding a boolean primarily for Arrow's use -- I eventually decided to include the new boolean in the output.  But I could be convinced to remove it there.  Either way we should probably be consistent.

File: cpp/src/parquet/arrow/arrow-schema-test.cc [hunk 2]
```diff
@@ -887,14 +891,15 @@ TEST_F(TestConvertArrowSchema, ArrowFields) {
...
+    arrow_fields.push_back(::arrow::field(c.name, c.datatype, false));
     parquet_fields.push_back(PrimitiveNode::Make(c.name, Repetition::REQUIRED,
                                                  c.logical_type, c.physical_type,
                                                  c.physical_length));
   }
 
   ASSERT_OK(ConvertSchema(arrow_fields));
-  ASSERT_NO_FATAL_FAILURE(CheckFlatSchema(parquet_fields));
+  CheckFlatSchema(parquet_fields);
+  // ASSERT_NO_FATAL_FAILURE();
```
Comment (wesm) [inline cpp/src/parquet/arrow/arrow-schema-test.cc]: I need to revert this

**Comment by wesm** [comment]
@tpboudreau yes, I'll add it to the `ToString` output -- I did initially and then removed it because it was annoying to adjust the unit tests =) But it makes debugging easier. I'll do that now so we can get this merged and then you can rebase your PR

**Comment by wesm** [comment]
Here's a build on my fork https://travis-ci.org/wesm/arrow/builds/557891740. The Travis CI failure appears unrelated. Will merge

**Pull request closed without merging**